src/scala/com/github/johnynek/bazel_deps/Normalizer.scala

```diff
@@ -86,9 +86,23 @@ object Normalizer {
           roots.contains(MavenCoordinate(node, v))
         }
         pickCanonical(node, rootVersion, dups, vcf) match {
-          case Validated.Valid(m) =>
-            val newItems = items.map { case (p, _) => (p, Right(m.version)) }
-            table.updated(node, newItems)
+          case Validated.Valid(canonical) =>
+            val newItems = items.map { case (p, _) => (p, Right(canonical.version)) }
+            table
+              // fix version of current entry
+              .updated(node, newItems)
+              // update all children
+              .mapValues(_.filter { case (parent, _) =>
```

Comment:
the intent seems good, but I think this algorithm is quadratic in the number of nodes.

Going from O(N) to O(N^2) here could be a deal breaker for many repos.

I think we could keep it O(N * E) (where E is the average number of edges nodes) if we use the `Graph[N, E]` data structure which can look up quickly edges that have a given source or destination. The input `graph` starts with that, but we may need to update that `graph` as we modify to use this algorithm.

I'm happy to merge something like this with some tests and as long we don't think we have any quadratic behavior.

Thanks for working on this.